obal/tito_extensions/git_source_strategy.py

```diff
@@ -0,0 +1,183 @@
+                    os.path.join(self.builder.rpmbuild_sourcedir, os.path.basename(s)))
+        print("  %s.spec" % self.builder.project_name)
+
+        replacements = []
+        src_files = run_command("find %s -type f" %
+              os.path.join(self.builder.rpmbuild_sourcedir, 'archive')).split("\n")
+
+        def filter_archives(path):
+            base_name = os.path.basename(path)
+            return ".tar" in base_name
+
+        for i, s in enumerate(filter(filter_archives, src_files)):
+            base_name = os.path.basename(s)
+            debug("Downloaded file %s" % base_name)
+
+            dest_filepath = os.path.join(self.builder.rpmbuild_sourcedir,
+                    base_name)
+            shutil.move(s, dest_filepath)
+            self.sources.append(dest_filepath)
+
+            # Add a line to replace in the spec for each source:
+            source_regex = re.compile("^(source%s:\s*)(.+)$" % i, re.IGNORECASE)
+            new_line = "Source%s: %s\n" % (i, base_name)
+            replacements.append((source_regex, new_line))
+
+        # Replace version in spec:
+        version_regex = re.compile("^(version:\s*)(.+)$", re.IGNORECASE)
+        self.version = self._get_version()
+        print("Building version: %s" % self.version)
+        replacements.append((version_regex, "Version: %s\n" % self.version))
+        self.replace_in_spec(replacements)
+
+        rel_date = datetime.utcnow().strftime("%Y%m%d%H%M")
+        self.release = rel_date + gitrev
+        print("Building release: %s" % self.release)
+        run_command("sed -i '/^Release:/ s/%%/.%s%%/' %s" % (self.release, self.spec_file))
+
+
+    """
+    Generates the source file from a local repo, useful for generating
+    packages of custom branches.
+    It first copies source files from git, then copies the source file
+    over the top, so they're merged (patches etc can then be stored
+    in git).
+    Takes the following arguments:
+      source_dir: path to repo ("~/foreman")
+    """
+    def _fetch_local(self):
+
+        source_dir = os.path.expanduser(self.builder.args['source_dir'][0])
+
+        version_regex = re.compile("^(version:\s*)(.+)$", re.IGNORECASE)
+
+        with open(self.spec_file, 'r') as spec:
+            for line in spec.readlines():
+                match = version_regex.match(line)
+                if match:
+                    version = match.group(2)
+
+        if not version:
+            error_out("Version not found in spec")
+
+        old_dir = os.getcwd()
+        os.chdir(source_dir)
+
+        fetchdir = os.path.join(self.builder.rpmbuild_sourcedir, 'archive')
+        if not os.path.exists(fetchdir):
+          os.mkdir(fetchdir)
+
+        arch_prefix = "-".join([self.builder.project_name, version])
+
+        # git archive --format=tar.gz --prefix=pulp-2.15.0/ master > pulp-2.15.0.tar.gz
+
+        with open("./%s.tar.gz" % arch_prefix, "w+") as archive:
```

Comment:
Foreman typically uses `.bz2`. The `.tar.gz` is also repeated a few times. Perhaps this should be its own separate variable? The `./` is also not needed.

Reply:
This is very dependent on what git archive supports as the "format" call.  I.E. you can't specify bz2, as git archive only supports a limited amount of formats.  (tar, tgz, tar.gz, zip).  This is explicitly utilizing git in a git checkout to generate the archive.

As far as `./` , while it is not needed, it is a little more visually explicit in what is going on.  Can change it though if it's strongly preferred.